agent/xds/routes.go

```diff
@@ -37,55 +37,95 @@ func routesFromSnapshotConnectProxy(cfgSnap *proxycfg.ConfigSnapshot) ([]proto.M
 		return nil, errors.New("nil config given")
 	}
 
-	return routesFromUpstreams(cfgSnap.ConnectProxy.ConfigSnapshotUpstreams, cfgSnap.Proxy.Upstreams)
-}
-
-// routesFromSnapshotIngressGateway returns the xDS API representation of the
-// "routes" in the snapshot.
-func routesFromSnapshotIngressGateway(cfgSnap *proxycfg.ConfigSnapshot) ([]proto.Message, error) {
-	if cfgSnap == nil {
-		return nil, errors.New("nil config given")
-	}
-
-	return routesFromUpstreams(cfgSnap.IngressGateway.ConfigSnapshotUpstreams, cfgSnap.IngressGateway.Upstreams)
-}
-
-func routesFromUpstreams(snap proxycfg.ConfigSnapshotUpstreams, upstreams structs.Upstreams) ([]proto.Message, error) {
 	var resources []proto.Message
-
-	for _, u := range upstreams {
+	for _, u := range cfgSnap.Proxy.Upstreams {
 		upstreamID := u.Identifier()
 
 		var chain *structs.CompiledDiscoveryChain
 		if u.DestinationType != structs.UpstreamDestTypePreparedQuery {
-			chain = snap.DiscoveryChain[upstreamID]
+			chain = cfgSnap.ConnectProxy.DiscoveryChain[upstreamID]
 		}
 
 		if chain == nil || chain.IsDefault() {
 			// TODO(rb): make this do the old school stuff too
 		} else {
-			upstreamRoute, err := makeUpstreamRouteForDiscoveryChain(&u, chain)
+			virtualHost, err := makeUpstreamRouteForDiscoveryChain(upstreamID, chain, []string{"*"})
 			if err != nil {
 				return nil, err
 			}
-			if upstreamRoute != nil {
-				resources = append(resources, upstreamRoute)
+
+			route := &envoy.RouteConfiguration{
+				Name:             upstreamID,
+				VirtualHosts:     []envoyroute.VirtualHost{*virtualHost},
+				ValidateClusters: makeBoolValue(true),
 			}
+			resources = append(resources, route)
 		}
 	}
 
 	// TODO(rb): make sure we don't generate an empty result
-
 	return resources, nil
 }
 
+// routesFromSnapshotIngressGateway returns the xDS API representation of the
+// "routes" in the snapshot.
+func routesFromSnapshotIngressGateway(cfgSnap *proxycfg.ConfigSnapshot) ([]proto.Message, error) {
+	if cfgSnap == nil {
+		return nil, errors.New("nil config given")
+	}
+
+	var result []proto.Message
+	for listenerKey, upstreams := range cfgSnap.IngressGateway.Upstreams {
+		// Do not create any route configuration for TCP listeners
+		if listenerKey.Protocol == "tcp" {
+			continue
+		}
+
+		upstreamRoute := &envoy.RouteConfiguration{
+			Name: listenerKey.RouteName(),
+			// ValidateClusters defaults to true when defined statically and false
+			// when done via RDS. Re-set the sane value of true to prevent
+			// null-routing traffic.
+			ValidateClusters: makeBoolValue(true),
+		}
+		for _, u := range upstreams {
+			upstreamID := u.Identifier()
+			chain := cfgSnap.IngressGateway.DiscoveryChain[upstreamID]
+			if chain != nil {
```

Comment:
Minor, but maybe invert this condition to a guard to reduce the indentation?

```go
if chain == nil {
    continue
}
```

Reply:
Done!